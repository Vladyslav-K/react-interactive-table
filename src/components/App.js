import React from "react";
import Container from "../components/container"
import Table from "../components/table";
import ButtonAddColumn from "../components/button-add-column";
import ButtonAddRow from "../components/button-add-row";

function App() {
  return (
    <Container>
      <Table />
      <ButtonAddColumn />
      <ButtonAddRow />
    </Container>
  );
}

export default App;
