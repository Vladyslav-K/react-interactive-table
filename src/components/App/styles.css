table {
  border: 1px solid #48aae6;
  cursor: grab;
}

td {
  background-color: #48aae6;
  height: 50px;
  width: 50px;
}

.error {
  font-size: 14px;
  text-align: center;
  padding: 10px;
  border: 2px solid red;
}

.container {
  display: inline-block;
  position: relative;
  left: 0;
  top: 0;
  margin: 65px;
}

.button {
  position: absolute;
  color: #ffffff;
  margin: 2px;
  height: 50px;
  width: 50px;
  font-size: 21px;
  font-weight: bold;
  text-align: center;
  transition-duration: 0.6s;
  opacity: 1;
  cursor: pointer;
}

.button:hover {
  opacity: 0.8;
  transition-duration: 0.5s;
}

.add {
  background-color: #f3a500;
  line-height: 50px;
}

.add-column {
  left: 100%;
  top: 1px;
}

.add-row {
  top: 100%;
  left: 1px;
}

.remove {
  background-color: #b20000;
  line-height: 40px;
  display: block;
  visibility: hidden;
}

.remove-column {
  bottom: 100%;
  margin: 2px 1px;
}

.remove-row {
  right: 100%;
  margin: 1px 2px;
}
