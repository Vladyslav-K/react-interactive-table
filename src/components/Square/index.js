import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const Wrapper = styled.div`
  display: flex;
  justify-content: center;

  user-select: none;
`;

const Container = styled.div`
  position: absolute;
  user-select: none;
`;

const Table = styled.table`
  border: 1px solid #48aae6;
  cursor: grab;
  user-select: none;
`;

const Cell = styled.td`
  height: ${({ cellSize }) => cellSize}px;
  min-width: ${({ cellSize }) => cellSize}px;
  min-height: ${({ cellSize }) => cellSize}px;

  padding: 0;

  background-color: #48aae6;

  user-select: none;
`;

const Button = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  position: absolute;

  height: ${({ cellSize }) => cellSize}px;
  width: ${({ cellSize }) => cellSize}px;
  margin: 2px;

  color: #ffffff;

  font-size: ${({ cellSize }) => cellSize / 2}px;
  font-weight: bold;
  text-align: center;

  opacity: 1;

  transition-duration: 0.6s;
  cursor: pointer;

  user-select: none;

  &:hover {
    opacity: 0.8;
    transition-duration: 0.5s;
  }
`;

const AddButtons = styled(Button)`
  background-color: #f3a500;
`;

const RemoveButtons = styled(Button)`
  visibility: hidden;
  opacity: 0;

  background-color: #b20000;
`;

const AddRowButton = styled(AddButtons)`
  top: 100%;
  left: 1px;
`;

const AddColumnButton = styled(AddButtons)`
  top: 1px;
  left: 100%;
`;

const RemoveRowButton = styled(RemoveButtons)`
  visibility: ${({ buttonsVisible }) =>
    buttonsVisible === true ? "visible" : "hidden"};
  opacity: ${({ buttonsVisible }) => (buttonsVisible === true ? 1 : 0)};

  top: ${({ cellSize, currentRowIndex }) =>
    cellSize * currentRowIndex + 2 * currentRowIndex}px;
  right: 100%;
  margin: 3px 2px;
`;

const RemoveColumnButton = styled(RemoveButtons)`
  visibility: ${({ buttonsVisible }) =>
    buttonsVisible === true ? "visible" : "hidden"};
  opacity: ${({ buttonsVisible }) => (buttonsVisible === true ? 1 : 0)};

  bottom: 100%;
  left: ${({ cellSize, currentCellIndex }) =>
    cellSize * currentCellIndex + 2 * currentCellIndex}px;
  margin: 2px 3px;
`;

export default class Square extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      rows: [],
      columns: [],

      currentCellIndex: 0,
      currentRowIndex: 0,

      buttonsVisible: false,
      removeRowButtonDisplay: true,
      removeColumnButtonDisplay: true,

      dragging: false
    };

    this.containerRef = React.createRef();

    this.uniqueKey = 0;

    this.offsetX = 0;
    this.offsetY = 0;
  }

  componentDidMount() {
    this.createSquare();
    document.addEventListener("mousemove", event => this.onDragging(event));
    document.addEventListener("mouseup", () => this.onDragEnd());
  }

  createSquare = () => {
    const newRows = [];
    const newColumns = [];

    const { initialHeight, initialWidth } = this.props;

    for (let cells = 0; cells < initialWidth; cells++) {
      newColumns[cells] = this.uniqueKey++;
    }

    for (let rows = 0; rows < initialHeight; rows++) {
      newRows[rows] = this.uniqueKey++;
    }

    this.setState({
      rows: newRows,
      columns: newColumns
    });
  };

  onDragStart = ({ clientX, clientY }) => {
    this.offsetX =
      clientX - this.containerRef.current.getBoundingClientRect().left;
    this.offsetY =
      clientY - this.containerRef.current.getBoundingClientRect().top;

    this.setState({
      dragging: true
    });
  };

  onDragging = ({ pageX, pageY }) => {
    const { dragging } = this.state;

    if (dragging) {
      this.containerRef.current.style.left = pageX - this.offsetX + "px";
      this.containerRef.current.style.top = pageY - this.offsetY + "px";
    }
  };

  onDragEnd = () => {
    this.setState({
      dragging: false
    });
  };

  movingButtons = ({ target }) => {
    if (target.tagName === "TD") {
      this.setState({
        currentCellIndex: target.cellIndex,
        currentRowIndex: target.parentNode.rowIndex
      });
    }
  };

  createColumn = () => {
    const { columns } = this.state;
    const cloneColumns = [...columns, this.uniqueKey++];

    this.setState({
      columns: cloneColumns
    });
  };

  createRow = () => {
    const { rows } = this.state;
    const cloneRows = [...rows, this.uniqueKey++];

    this.setState({
      rows: cloneRows
    });
  };

  deleteColumn = () => {
    const { currentCellIndex, columns } = this.state;
    const lastCellIndex = columns.length - 1;

    if (columns.length > 1) {
      this.setState({
        columns: columns.filter((column, index) => index !== currentCellIndex)
      });
    }

    if (currentCellIndex === lastCellIndex) {
      this.setState(prevState => ({
        currentCellIndex: prevState.currentCellIndex - 1
      }));
    }

    if (lastCellIndex <= 1) {
      this.setState({
        removeColumnButtonDisplay: false
      });
    }
  };

  deleteRow = () => {
    const { currentRowIndex, rows } = this.state;
    const lastRowIndex = rows.length - 1;

    if (rows.length > 1) {
      this.setState({
        rows: rows.filter((row, index) => index !== currentRowIndex)
      });
    }

    if (currentRowIndex === lastRowIndex) {
      this.setState(prevState => ({
        currentRowIndex: prevState.currentRowIndex - 1
      }));
    }

    if (lastRowIndex <= 1) {
      this.setState({
        removeRowButtonDisplay: false
      });
    }
  };

  showButtons = () => {
    const { rows, columns } = this.state;

    if (rows.length > 1) {
      this.setState({
        buttonsVisible: true,
        removeRowButtonDisplay: true
      });
    }

    if (columns.length > 1) {
      this.setState({
        buttonsVisible: true,
        removeColumnButtonDisplay: true
      });
    }
  };

  hideButtons = () => {
    this.setState({
      buttonsVisible: false
    });
  };

  render() {
    const { cellSize } = this.props;

    const {
      rows,
      columns,
      buttonsVisible,
      removeRowButtonDisplay,
      removeColumnButtonDisplay,
      currentRowIndex,
      currentCellIndex
    } = this.state;

    return (
      <Wrapper>
        <Container ref={this.containerRef} onMouseOver={this.movingButtons}>
          <div onMouseEnter={this.showButtons} onMouseLeave={this.hideButtons}>
            <Table onMouseDown={this.onDragStart} onDragStart={() => false}>
              <tbody>
                {rows.map(row => (
                  <tr key={`row-${row}`} data-test="row">
                    {columns.map(cell => (
                      <Cell
                        data-test="cell"
                        key={`cell-${cell}`}
                        cellSize={cellSize}
                      />
                    ))}
                  </tr>
                ))}
              </tbody>
            </Table>

            {removeRowButtonDisplay && (
              <RemoveRowButton
                data-test="remove-row-button"
                onClick={this.deleteRow}
                cellSize={cellSize}
                currentRowIndex={currentRowIndex}
                buttonsVisible={buttonsVisible}
              >
                -
              </RemoveRowButton>
            )}

            {removeColumnButtonDisplay && (
              <RemoveColumnButton
                data-test="remove-column-button"
                onClick={this.deleteColumn}
                cellSize={cellSize}
                currentCellIndex={currentCellIndex}
                buttonsVisible={buttonsVisible}
              >
                -
              </RemoveColumnButton>
            )}
          </div>
          <AddRowButton
            data-test="add-row-button"
            cellSize={cellSize}
            onClick={this.createRow}
          >
            +
          </AddRowButton>
          <AddColumnButton
            data-test="add-column-button"
            cellSize={cellSize}
            onClick={this.createColumn}
          >
            +
          </AddColumnButton>
        </Container>
      </Wrapper>
    );
  }
}

Square.propTypes = {
  cellSize: PropTypes.number.isRequired,
  initialHeight: PropTypes.number.isRequired,
  initialWidth: PropTypes.number.isRequired
};
