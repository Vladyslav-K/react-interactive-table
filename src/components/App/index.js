import React from "react";
import Square from "../Square";
import styled from "styled-components";

const GlobalContainer = styled.div`
  display: flex;
  justify-content: space-around;
  padding-top: 150px;
`;

function App() {
  return (
    <GlobalContainer>
      <Square initialWidth={4} initialHeight={4} cellSize={50} />
      <Square initialWidth={4} initialHeight={4} cellSize={50} />
    </GlobalContainer>
  );
}

export default App;
