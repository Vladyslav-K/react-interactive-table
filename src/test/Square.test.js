import React from "react";
import { shallow } from "enzyme";

import App from "../components/App";
import Square from "../components/Square";

describe("Render", () => {
  it("Component APP rendered", () => {
    const AppShallow = shallow(<App />);

    expect(AppShallow).toMatchSnapshot();
  });

  it("Component Square rendered", () => {
    const SquareShallow = shallow(
      <Square initialHeight={4} initialWidth={4} cellSize={50} />
    );

    expect(SquareShallow).toMatchSnapshot();
  });
});

describe("Square", () => {
  it("Correct initial height", () => {
    const SquareShallow = shallow(<Square />);
    const AppShallow = shallow(<App />);
    const { initialHeight } = AppShallow.props();

    expect(SquareShallow.find('[data-test="row"]')).toHaveLength(initialHeight);
  });

  it("Correct initial width", () => {
    const SquareShallow = shallow(<Square />);
    const AppShallow = shallow(<App />);
    const { initialWidth, initialHeight } = AppShallow.props();

    expect(SquareShallow.find('[data-test="cell"]')).toHaveLength(
      initialWidth * initialHeight
    );
  });

  it("Clicking on the addRowButton creates a row", () => {
    const SquareShallow = shallow(<Square />);
    const AppShallow = shallow(<App />);
    const { initialHeight } = AppShallow.props();

    SquareShallow.find('[data-test="add-row-button"]').simulate("click");
    expect(SquareShallow.find('[data-test="row"]')).toHaveLength(
      initialHeight + 1
    );
  });

  it("Clicking on the addColumnButton creates a column", () => {
    const SquareShallow = shallow(<Square />);
    const AppShallow = shallow(<App />);
    const { initialWidth, initialHeight } = AppShallow.props();

    SquareShallow.find('[data-test="add-column-button"]').simulate("click");

    expect(SquareShallow.find('[data-test="cell"]')).toHaveLength(
      (initialWidth + 1) * initialHeight
    );
  });

  it("Clicking on the removeRowButton deletes a row", () => {
    const SquareShallow = shallow(<Square />);
    const AppShallow = shallow(<App />);
    const { initialHeight } = AppShallow.props();

    SquareShallow.find('[data-test="remove-row-button"]').simulate("click");

    expect(SquareShallow.find('[data-test="row"]')).toHaveLength(
      initialHeight - 1
    );
  });

  it("Clicking on the removeColumnButton deletes a column", () => {
    const SquareShallow = shallow(<Square />);
    const AppShallow = shallow(<App />);
    const { initialWidth, initialHeight } = AppShallow.props();

    SquareShallow.find('[data-test="remove-column-button"]').simulate("click");

    expect(SquareShallow.find('[data-test="cell"]')).toHaveLength(
      initialHeight * initialWidth - initialWidth
    );
  });

  it("removeRowButton does not delete row if last", () => {
    const OneCellSquare = shallow(
      <Square initialHeight={1} initialWidth={1} cellSize={50} />
    );

    OneCellSquare.find('[data-test="remove-row-button"]').simulate("click");

    expect(OneCellSquare.find('[data-test="row"]')).toHaveLength(1);
  });

  it("removeColumnButton does not delete column if last", () => {
    const OneCellSquare = shallow(
      <Square initialHeight={1} initialWidth={1} cellSize={50} />
    );

    OneCellSquare.find('[data-test="remove-column-button"]').simulate("click");

    expect(OneCellSquare.find('[data-test="cell"]')).toHaveLength(1);
  });
});
