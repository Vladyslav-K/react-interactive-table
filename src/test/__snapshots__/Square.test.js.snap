// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Component APP rendered 1`] = `
<styled.div>
  <Square
    cellSize={50}
    initialHeight={4}
    initialWidth={4}
  />
  <Square
    cellSize={50}
    initialHeight={4}
    initialWidth={4}
  />
</styled.div>
`;

exports[`Render Component Square rendered 1`] = `
<styled.div>
  <styled.div
    onMouseOver={[Function]}
  >
    <div
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <styled.table
        onDragStart={[Function]}
        onMouseDown={[Function]}
      >
        <tbody>
          <tr
            data-test="row"
            key="row-4"
          >
            <styled.td
              cellSize={50}
              data-test="cell"
              key="cell-0"
            />
            <styled.td
              cellSize={50}
              data-test="cell"
              key="cell-1"
            />
            <styled.td
              cellSize={50}
              data-test="cell"
              key="cell-2"
            />
            <styled.td
              cellSize={50}
              data-test="cell"
              key="cell-3"
            />
          </tr>
          <tr
            data-test="row"
            key="row-5"
          >
            <styled.td
              cellSize={50}
              data-test="cell"
              key="cell-0"
            />
            <styled.td
              cellSize={50}
              data-test="cell"
              key="cell-1"
            />
            <styled.td
              cellSize={50}
              data-test="cell"
              key="cell-2"
            />
            <styled.td
              cellSize={50}
              data-test="cell"
              key="cell-3"
            />
          </tr>
          <tr
            data-test="row"
            key="row-6"
          >
            <styled.td
              cellSize={50}
              data-test="cell"
              key="cell-0"
            />
            <styled.td
              cellSize={50}
              data-test="cell"
              key="cell-1"
            />
            <styled.td
              cellSize={50}
              data-test="cell"
              key="cell-2"
            />
            <styled.td
              cellSize={50}
              data-test="cell"
              key="cell-3"
            />
          </tr>
          <tr
            data-test="row"
            key="row-7"
          >
            <styled.td
              cellSize={50}
              data-test="cell"
              key="cell-0"
            />
            <styled.td
              cellSize={50}
              data-test="cell"
              key="cell-1"
            />
            <styled.td
              cellSize={50}
              data-test="cell"
              key="cell-2"
            />
            <styled.td
              cellSize={50}
              data-test="cell"
              key="cell-3"
            />
          </tr>
        </tbody>
      </styled.table>
      <Styled(Styled(styled.div))
        buttonsVisible={false}
        cellSize={50}
        currentRowIndex={0}
        data-test="remove-row-button"
        onClick={[Function]}
      >
        -
      </Styled(Styled(styled.div))>
      <Styled(Styled(styled.div))
        buttonsVisible={false}
        cellSize={50}
        currentCellIndex={0}
        data-test="remove-column-button"
        onClick={[Function]}
      >
        -
      </Styled(Styled(styled.div))>
    </div>
    <Styled(Styled(styled.div))
      cellSize={50}
      data-test="add-row-button"
      onClick={[Function]}
    >
      +
    </Styled(Styled(styled.div))>
    <Styled(Styled(styled.div))
      cellSize={50}
      data-test="add-column-button"
      onClick={[Function]}
    >
      +
    </Styled(Styled(styled.div))>
  </styled.div>
</styled.div>
`;
